File: Practica4_REGISTROS.py
# Importa el módulo csv, que permite leer y escribir archivos en formato CSV
import csv

# Importa el módulo os, que proporciona funciones para interactuar con el sistema operativo.
import os


# -----------------------------------------------------------------------------
# CLASE ESTUDIANTE
# -----------------------------------------------------------------------------
# Esta clase representa un nodo de una lista enlazada.
# Cada objeto Estudiante almacena los datos de un estudiante y una referencia
# al siguiente nodo de la lista.
class Estudiante:

    # Método constructor. Se ejecuta automáticamente al crear un objeto.
    def __init__(self, id, nombre, correo, escuela, anio_ingreso):

        # Identificador único del estudiante.
        self.id = id

        # Nombre completo del estudiante.
        self.nombre = nombre

        # Correo electrónico.
        self.correo = correo

        # Escuela profesional o carrera.
        self.escuela = escuela

        # Año de ingreso a la institución.
        self.anio_ingreso = anio_ingreso

        # Referencia al siguiente nodo de la lista enlazada.
        # Inicialmente no apunta a ningún nodo.
        self.siguiente = None


# -----------------------------------------------------------------------------
# CLASE LISTAESTUDIANTES
# -----------------------------------------------------------------------------
# Implementa una lista enlazada simple para almacenar objetos Estudiante.
# Incluye operaciones de agregar, mostrar, buscar, modificar, eliminar,
# guardar y cargar desde archivo.
class ListaEstudiantes:

    # Constructor de la clase.
    def __init__(self):

        # Referencia al primer nodo de la lista.
        # Si la lista está vacía, su valor es None.
        self.cabeza = None

        # Variable que almacena el próximo ID disponible.
        self.siguiente_id = 1


    # -------------------------------------------------------------------------
    # AGREGAR ESTUDIANTE
    # -------------------------------------------------------------------------
    # Crea un nuevo nodo y lo inserta al final de la lista.
    def agregar_estudiante(self, nombre, correo, escuela, anio_ingreso):

        # Crea un nuevo objeto Estudiante usando el ID actual.
        nuevo = Estudiante(self.siguiente_id, nombre, correo, escuela, anio_ingreso)

        # Incrementa el contador para que el próximo estudiante tenga un ID único.
        self.siguiente_id += 1

        # Si la lista está vacía.
        if self.cabeza is None:

            # El nuevo nodo se convierte en la cabeza.
            self.cabeza = nuevo

        else:

            # Comienza desde el primer nodo.
            actual = self.cabeza

            # Recorre la lista hasta llegar al último nodo.
            while actual.siguiente is not None:
                actual = actual.siguiente

            # Conecta el último nodo con el nuevo.
            actual.siguiente = nuevo

        # Muestra un mensaje de confirmación.
        print(f"Estudiante agregado con éxito. ID asignado: {nuevo.id}")


    # -------------------------------------------------------------------------
    # MOSTRAR TODOS LOS ESTUDIANTES
    # -------------------------------------------------------------------------
    # Recorre la lista e imprime la información de cada estudiante.
    def mostrar_estudiantes(self):

        # Si la lista está vacía.
        if self.cabeza is None:
            print("No hay estudiantes registrados.")
            return

        # Apunta al primer nodo.
        actual = self.cabeza

        # Encabezado.
        print("--- Lista de Estudiantes ---")

        # Recorre la lista hasta llegar al final.
        while actual is not None:

            # Muestra los datos del nodo actual.
            print(f"ID: {actual.id}")
            print(f"Nombre: {actual.nombre}")
            print(f"Correo: {actual.correo}")
            print(f"Escuela: {actual.escuela}")
            print(f"Año de Ingreso: {actual.anio_ingreso}")
            print("---------------------------")

            # Avanza al siguiente nodo.
            actual = actual.siguiente


    # -------------------------------------------------------------------------
    # BUSCAR POR ID
    # -------------------------------------------------------------------------
    # Busca un estudiante utilizando su identificador único.
    def buscar_por_id(self, id_buscar):

        # Inicia desde la cabeza.
        actual = self.cabeza

        # Recorre todos los nodos.
        while actual is not None:

            # Si el ID coincide.
            if actual.id == id_buscar:

                # Muestra la información del estudiante.
                print("--- Estudiante Encontrado ---")
                print(f"ID: {actual.id}")
                print(f"Nombre: {actual.nombre}")
                print(f"Correo: {actual.correo}")
                print(f"Escuela: {actual.escuela}")
                print(f"Año de Ingreso: {actual.anio_ingreso}")

                # Termina la función inmediatamente.
                return

            # Continúa con el siguiente nodo.
            actual = actual.siguiente

        # Si no se encontró ningún estudiante con ese ID.
        print(f"Estudiante con ID {id_buscar} no encontrado.")


    # -------------------------------------------------------------------------
    # BUSCAR POR NOMBRE
    # -------------------------------------------------------------------------
    # Busca estudiantes cuyo nombre contenga el texto ingresado.
    # La búsqueda no distingue mayúsculas y minúsculas.
    def buscar_por_nombre(self, nombre_buscar):

        # Inicia desde la cabeza.
        actual = self.cabeza

        # Bandera para saber si se encontró al menos una coincidencia.
        encontrado = False

        # Recorre la lista.
        while actual is not None:

            # Convierte ambos textos a minúsculas y verifica si uno está dentro del otro.
            if nombre_buscar.lower() in actual.nombre.lower():

                # Muestra los datos del estudiante encontrado.
                print("--- Estudiante Encontrado ---")
                print(f"ID: {actual.id}")
                print(f"Nombre: {actual.nombre}")
                print(f"Correo: {actual.correo}")
                print(f"Escuela: {actual.escuela}")
                print(f"Año de Ingreso: {actual.anio_ingreso}")

                # Marca que sí hubo coincidencias.
                encontrado = True

            # Avanza al siguiente nodo.
            actual = actual.siguiente

        # Si no se encontró ninguna coincidencia.
        if not encontrado:
            print(f"Ningún estudiante encontrado con el nombre: {nombre_buscar}")


    # -------------------------------------------------------------------------
    # MODIFICAR ESTUDIANTE
    # -------------------------------------------------------------------------
    # Permite actualizar uno o varios campos de un estudiante.
    def modificar_estudiante(self, id_modificar):

        # Comienza desde la cabeza.
        actual = self.cabeza

        # Recorre toda la lista.
        while actual is not None:

            # Si encuentra el ID solicitado.
            if actual.id == id_modificar:

                print(f"--- Modificando Estudiante ID {id_modificar} ---")
                print("Dejar en blanco (o presionar Enter) para no modificar el campo.")

                # Solicita nuevo nombre.
                nombre = input(f"Nombre actual ({actual.nombre}): ")

                # Si el usuario escribió algo distinto de espacios.
                if nombre.strip(): actual.nombre = nombre

                # Solicita nuevo correo.
                correo = input(f"Correo actual ({actual.correo}): ")

                # Actualiza si se ingresó un valor.
                if correo.strip(): actual.correo = correo

                # Solicita nueva escuela.
                escuela = input(f"Escuela actual ({actual.escuela}): ")

                # Actualiza si corresponde.
                if escuela.strip(): actual.escuela = escuela

                # Solicita nuevo año de ingreso.
                anio = input(f"Año de Ingreso actual ({actual.anio_ingreso}): ")

                # Si se ingresó algún valor.
                if anio.strip(): 
                    try:
                        # Convierte la cadena a entero.
                        actual.anio_ingreso = int(anio)
                    except ValueError:
                        # Si la conversión falla.
                        print("Entrada inválida. El año no se modificó.") 

                # Confirmación.
                print("Datos actualizados exitosamente.")
                return

            # Continúa con el siguiente nodo.
            actual = actual.siguiente

        # Si no se encontró el ID.
        print(f"Estudiante con ID {id_modificar} no encontrado.")


    # -------------------------------------------------------------------------
    # ELIMINAR ESTUDIANTE
    # -------------------------------------------------------------------------
    # Elimina un nodo de la lista según su ID.
    def eliminar_estudiante(self, id_eliminar):

        # Si la lista está vacía.
        if self.cabeza is None:
            print("La lista está vacía.")
            return

        # Caso especial: el nodo a eliminar es la cabeza.
        if self.cabeza.id == id_eliminar:

            # La cabeza pasa a ser el siguiente nodo.
            self.cabeza = self.cabeza.siguiente

            # Confirmación.
            print("Estudiante eliminado exitosamente.")
            return

        # Comienza desde la cabeza.
        actual = self.cabeza

        # Busca el nodo anterior al que se desea eliminar.
        while actual.siguiente is not None and actual.siguiente.id != id_eliminar:
            actual = actual.siguiente

        # Si se encontró el nodo.
        if actual.siguiente is not None:

            # Se salta el nodo eliminado.
            actual.siguiente = actual.siguiente.siguiente

            # Confirmación.
            print("Estudiante eliminado exitosamente.")

        else:

            # Si no se encontró el ID.
            print(f"Estudiante con ID {id_eliminar} no encontrado.") 


    # -------------------------------------------------------------------------
    # GUARDAR EN ARCHIVO CSV
    # -------------------------------------------------------------------------
    # Escribe todos los estudiantes en un archivo CSV.
    def guardar_en_archivo(self, filename):

        # Abre el archivo en modo escritura ('w').
        # newline='' evita líneas en blanco extra.
        # encoding='utf-8' permite caracteres especiales.
        with open(filename, mode='w', newline='', encoding='utf-8') as archivo:

            # Crea un escritor CSV.
            writer = csv.writer(archivo)

            # Escribe la fila de encabezado.
            writer.writerow(['ID', 'Nombre', 'Correo', 'Escuela', 'Año_Ingreso'])

            # Comienza desde la cabeza.
            actual = self.cabeza

            # Recorre toda la lista.
            while actual is not None:

                # Escribe una fila con los datos del estudiante.
                writer.writerow([
                    actual.id,
                    actual.nombre,
                    actual.correo,
                    actual.escuela,
                    actual.anio_ingreso
                ])

                # Avanza al siguiente nodo.
                actual = actual.siguiente

        # Al salir del bloque with, el archivo se cierra automáticamente.
        print(f"Datos guardados en {filename} exitosamente.") 


    # -------------------------------------------------------------------------
    # CARGAR DESDE ARCHIVO CSV
    # -------------------------------------------------------------------------
    # Lee los datos almacenados y reconstruye la lista enlazada.
    def cargar_desde_archivo(self, filename):

        # Verifica si el archivo existe.
        if not os.path.exists(filename): 
            return

        # Reinicia la lista.
        self.cabeza = None

        # Referencia al último nodo insertado.
        ultimo = None

        # Guarda el ID máximo encontrado.
        max_id = 0

        # Abre el archivo en modo lectura.
        with open(filename, mode='r', newline='', encoding='utf-8') as archivo:

            # Crea un lector CSV.
            reader = csv.reader(archivo)

            try:
                # Omite la primera fila (encabezado).
                next(reader) # Saltar el encabezado
            except StopIteration:
                # Si el archivo está vacío.
                return

            # Recorre cada fila del archivo.
            for row in reader:

                # Verifica que tenga exactamente 5 columnas.
                if len(row) == 5:

                    # Crea un nuevo objeto Estudiante con los datos del CSV.
                    nuevo = Estudiante(
                        int(row[0]),
                        row[1],
                        row[2],
                        row[3],
                        int(row[4])
                    )

                    # Actualiza el ID máximo encontrado.
                    if nuevo.id > max_id:
                        max_id = nuevo.id

                    # Inserta el nodo en la lista.
                    if self.cabeza is None:
                        self.cabeza = nuevo
                        ultimo = nuevo
                    else:
                        ultimo.siguiente = nuevo
                        ultimo = nuevo

        # Ajusta el siguiente ID disponible.
        self.siguiente_id = max_id + 1

        # Confirmación.
        print(f"Datos cargados desde {filename} exitosamente.") 


# -----------------------------------------------------------------------------
# FUNCIÓN MOSTRAR MENÚ
# -----------------------------------------------------------------------------
# Imprime el menú principal y retorna la opción elegida por el usuario.
def mostrar_menu(): 

    print("\n=== SISTEMA DE REGISTRO DE ESTUDIANTES ===")
    print("1. Agregar nuevo estudiante")
    print("2. Mostrar todos los estudiantes")
    print("3. Buscar estudiante por ID o nombre")
    print("4. Modificar datos de un estudiante")
    print("5. Eliminar un estudiante")
    print("6. Guardar los datos en archivo .csv")
    print("7. Cargar los datos desde archivo")
    print("8. Salir")

    # Solicita la opción al usuario.
    opcion = input("Seleccione una opción: ")

    # Retorna la opción como cadena.
    return opcion


# -----------------------------------------------------------------------------
# FUNCIÓN PRINCIPAL
# -----------------------------------------------------------------------------
# Controla el flujo general del programa.
def main(): 

    # Crea el objeto que administrará la lista enlazada.
    lista = ListaEstudiantes()

    # Nombre del archivo de almacenamiento.
    filename = "registros.csv"

    # Carga automáticamente los datos al iniciar.
    lista.cargar_desde_archivo(filename)

    # Bucle infinito del menú.
    while True:

        # Muestra el menú y obtiene la opción.
        opcion = mostrar_menu()

        # ---------------------------------------------------------------------
        # OPCIÓN 1: AGREGAR ESTUDIANTE
        # ---------------------------------------------------------------------
        if opcion == '1':
            nombre = input("Ingrese nombre completo: ")
            correo = input("Ingrese correo electrónico: ")
            escuela = input("Ingrese escuela profesional: ")

            try:
                anio = int(input("Ingrese año de ingreso: "))
                lista.agregar_estudiante(nombre, correo, escuela, anio)
            except ValueError:
                print("El año de ingreso debe ser un número entero.") 

        # ---------------------------------------------------------------------
        # OPCIÓN 2: MOSTRAR
        # ---------------------------------------------------------------------
        elif opcion == '2':
            lista.mostrar_estudiantes()

        # ---------------------------------------------------------------------
        # OPCIÓN 3: BUSCAR
        # ---------------------------------------------------------------------
        elif opcion == '3':
            subopcion = input("1. Buscar por ID\n2. Buscar por Nombre\nSeleccione: ") 

            if subopcion == '1':
                try:
                    id_buscar = int(input("Ingrese ID a buscar: "))
                    lista.buscar_por_id(id_buscar)
                except ValueError:
                    print("Entrada inválida.") 

            elif subopcion == '2':
                nombre_buscar = input("Ingrese nombre a buscar: ")
                lista.buscar_por_nombre(nombre_buscar)

            else:
                print("Opción inválida.") 

        # ---------------------------------------------------------------------
        # OPCIÓN 4: MODIFICAR
        # ---------------------------------------------------------------------
        elif opcion == '4':
            try:
                id_modificar = int(input("Ingrese el ID del estudiante a modificar: "))
                lista.modificar_estudiante(id_modificar)
            except ValueError:
                print("Entrada inválida.") 

        # ---------------------------------------------------------------------
        # OPCIÓN 5: ELIMINAR
        # ---------------------------------------------------------------------
        elif opcion == '5':
            try:
                id_eliminar = int(input("Ingrese el ID del estudiante a eliminar: "))
                lista.eliminar_estudiante(id_eliminar)
            except ValueError:
                print("Entrada inválida.") 

        # ---------------------------------------------------------------------
        # OPCIÓN 6: GUARDAR
        # ---------------------------------------------------------------------
        elif opcion == '6':
            lista.guardar_en_archivo(filename)

        # ---------------------------------------------------------------------
        # OPCIÓN 7: CARGAR
        # ---------------------------------------------------------------------
        elif opcion == '7':
            lista.cargar_desde_archivo(filename)

        # ---------------------------------------------------------------------
        # OPCIÓN 8: SALIR
        # ---------------------------------------------------------------------
        elif opcion == '8':
            print("Guardando datos antes de salir...")
            lista.guardar_en_archivo(filename)
            print("Saliendo del programa...") 
            break

        # ---------------------------------------------------------------------
        # OPCIÓN INVÁLIDA
        # ---------------------------------------------------------------------
        else:
            print("Opción inválida. Intente nuevamente.") 


# -----------------------------------------------------------------------------
# PUNTO DE ENTRADA DEL PROGRAMA
# -----------------------------------------------------------------------------
# Esta condición garantiza que main() se ejecute solo si este archivo
# se ejecuta directamente, y no cuando se importa como módulo.
if __name__ == "__main__": 
    main()

File: Practica4_REGISTROS.cpp
#include <iostream> // Permite usar cout (mostrar en pantalla) y cin (leer desde teclado).
#include <fstream> // Permite usar ifstream (lectura) y ofstream (escritura).
#include <string> // Facilita el manejo de cadenas de texto.
#include <sstream> // Permite dividir y procesar cadenas como si fueran flujos de datos.

using namespace std; // Evita escribir std:: antes de cout, cin, string, etc.

//ESTRUCTURA ESTUDIANTE
// Define un nodo de la lista enlazada. // Cada nodo almacena los datos de un estudiante y un puntero al siguiente nodo
struct Estudiante {
    int id;  // Identificador único del estudiante.
    string nombre;          
    string correo;          
    string escuela;   // Escuela profesional.
    int anioIngreso;        
    
    Estudiante* siguiente; // Puntero al siguiente nodo de la lista. // Si es el último nodo, este puntero será nullptr.
};

// POO - Utilizamos el Paradigma de Programación Orientada a Objetos 
// CLASE ListaEstudiantes  // Administra la lista enlazada de estudiantes y todas las operaciones.
class ListaEstudiantes {
private:
    Estudiante* cabeza; // Puntero al primer nodo de la lista. // Si la lista está vacía, vale nullptr.
    int siguienteId; // Variable que guarda el próximo ID a asignar automáticamente.

public:
    // CONSTRUCTOR
    ListaEstudiantes() { // Inicializa una lista vacía.
        cabeza = nullptr;   // No hay nodos al inicio.
        siguienteId = 1;    // El primer estudiante tendrá ID 1.
    }

    // DESTRUCTOR
    ~ListaEstudiantes() {
        Estudiante* actual = cabeza;  // Comenzamos desde la cabeza.
        while (actual != nullptr) { // Recorremos toda la lista.
            Estudiante* aBorrar = actual;  // Guardamos el nodo actual para eliminarlo.
            actual = actual->siguiente;  // Avanzamos al siguiente nodo.
            delete aBorrar;  // Liberamos la memoria del nodo.
        }
    }

    // AGREGAR ESTUDIANTE
    void agregarEstudiante(string nombre, string correo,
                           string escuela, int anioIngreso) {

        Estudiante* nuevo = new Estudiante;  // Reservamos memoria para un nuevo nodo.
        nuevo->id = siguienteId++;  // Asignamos un ID único y luego incrementamos el contador.

        // Copiamos los datos recibidos.
        nuevo->nombre = nombre;
        nuevo->correo = correo;
        nuevo->escuela = escuela;
        nuevo->anioIngreso = anioIngreso;

        // Como será el último nodo, apunta a nullptr.
        nuevo->siguiente = nullptr;

        // Si la lista está vacía.
        if (cabeza == nullptr) {
            // El nuevo nodo se convierte en la cabeza.
            cabeza = nuevo;
        }
        else {
            // Recorremos hasta el último nodo.
            Estudiante* actual = cabeza;
            while (actual->siguiente != nullptr) {
                actual = actual->siguiente;
            }

            // Enlazamos el último nodo con el nuevo.
            actual->siguiente = nuevo;
        }

        // Confirmación al usuario.
        cout << "Estudiante agregado con exito. ID asignado: "
             << nuevo->id << endl;
    }

    // MOSTRAR TODOS LOS ESTUDIANTES
    void mostrarEstudiantes() {

        // Si la lista está vacía.
        if (cabeza == nullptr) {
            cout << "No hay estudiantes registrados." << endl;
            return;
        }

        // Empezamos desde el primer nodo.
        Estudiante* actual = cabeza;

        cout << "--- Lista de Estudiantes ---" << endl;

        // Recorremos toda la lista.
        while (actual != nullptr) {
            cout << "ID: " << actual->id << endl;
            cout << "Nombre: " << actual->nombre << endl;
            cout << "Correo: " << actual->correo << endl;
            cout << "Escuela: " << actual->escuela << endl;
            cout << "Anio de Ingreso: "
                 << actual->anioIngreso << endl;
            cout << "---------------------------" << endl;

            // Avanzamos al siguiente nodo.
            actual = actual->siguiente;
        }
    }

    // BUSCAR ESTUDIANTE POR ID
    void buscarEstudiante(int id) {

        // Iniciamos desde la cabeza.
        Estudiante* actual = cabeza;

        // Recorremos la lista.
        while (actual != nullptr) {

            // Si encontramos el ID.
            if (actual->id == id) {
                cout << "--- Estudiante Encontrado ---" << endl;
                cout << "ID: " << actual->id << endl;
                cout << "Nombre: " << actual->nombre << endl;
                cout << "Correo: " << actual->correo << endl;
                cout << "Escuela: " << actual->escuela << endl;
                cout << "Anio de Ingreso: "
                     << actual->anioIngreso << endl;

                // Terminamos la función.
                return;
            }

            // Avanzamos al siguiente nodo.
            actual = actual->siguiente;
        }

        // Si no se encontró ningún nodo con ese ID.
        cout << "Estudiante con ID " << id
             << " no encontrado." << endl;
    }

    // BUSCAR ESTUDIANTE POR NOMBRE
    void buscarEstudiante(string nombre) {

        // Comenzamos desde la cabeza.
        Estudiante* actual = cabeza;

        // Bandera para saber si hubo coincidencias.
        bool encontrado = false;

        // Recorremos toda la lista.
        while (actual != nullptr) {

            // find() busca una subcadena.
            // Si no la encuentra devuelve string::npos.
            if (actual->nombre.find(nombre) != string::npos) {

                cout << "--- Estudiante Encontrado ---" << endl;
                cout << "ID: " << actual->id << endl;
                cout << "Nombre: " << actual->nombre << endl;
                cout << "Correo: " << actual->correo << endl;
                cout << "Escuela: " << actual->escuela << endl;
                cout << "Anio de Ingreso: "
                     << actual->anioIngreso << endl;

                // Marcamos que sí se encontró al menos uno.
                encontrado = true;
            }

            // Avanzamos al siguiente nodo.
            actual = actual->siguiente;
        }

        // Si no hubo coincidencias.
        if (!encontrado) {
            cout << "Ningun estudiante encontrado con el nombre: "
                 << nombre << endl;
        }
    }

    // MODIFICAR ESTUDIANTE
    void modificarEstudiante(int id) {

        // Empezamos desde la cabeza.
        Estudiante* actual = cabeza;

        // Buscamos el ID.
        while (actual != nullptr) {

            if (actual->id == id) {
                cout << "--- Modificando Estudiante ID "
                     << id << " ---" << endl;

                cout << "Dejar en blanco (o presionar Enter) "
                     << "para no modificar el campo." << endl;

                // Variable temporal para leer cada campo.
                string entrada;

                // Limpiamos el buffer para que getline funcione correctamente.
                cin.ignore(10000, '\n');

                // ----------------- Nombre -----------------
                cout << "Nombre actual ("
                     << actual->nombre << "): ";
                getline(cin, entrada);

                // Si el usuario escribió algo, actualizamos.
                if (!entrada.empty())
                    actual->nombre = entrada;

                // ----------------- Correo -----------------
                cout << "Correo actual ("
                     << actual->correo << "): ";
                getline(cin, entrada);

                if (!entrada.empty())
                    actual->correo = entrada;

                // ----------------- Escuela -----------------
                cout << "Escuela actual ("
                     << actual->escuela << "): ";
                getline(cin, entrada);

                if (!entrada.empty())
                    actual->escuela = entrada;

                // ----------------- Año de ingreso -----------------
                cout << "Anio de Ingreso actual ("
                     << actual->anioIngreso << "): ";
                getline(cin, entrada);

                // Si se ingresó un valor.
                if (!entrada.empty()) {
                    try {
                        // Convertimos la cadena a entero.
                        actual->anioIngreso = stoi(entrada);
                    }
                    catch (...) {
                        // Si la conversión falla.
                        cout << "Entrada invalida. "
                             << "El anio no se modifico." << endl;
                    }
                }

                cout << "Datos actualizados exitosamente." << endl;
                return;
            }

            // Avanzamos al siguiente nodo.
            actual = actual->siguiente;
        }

        // Si no se encontró el ID.
        cout << "Estudiante con ID " << id
             << " no encontrado." << endl;
    }

    // ELIMINAR ESTUDIANTE
    void eliminarEstudiante(int id) {

        // Si la lista está vacía.
        if (cabeza == nullptr) {
            cout << "La lista esta vacia." << endl;
            return;
        }

        // Caso especial: el nodo a eliminar es la cabeza.
        if (cabeza->id == id) {
            Estudiante* aBorrar = cabeza;

            // La cabeza pasa a ser el segundo nodo.
            cabeza = cabeza->siguiente;

            // Liberamos memoria.
            delete aBorrar;

            cout << "Estudiante eliminado exitosamente." << endl;
            return;
        }

        // Recorremos hasta encontrar el nodo anterior al que se eliminará.
        Estudiante* actual = cabeza;

        while (actual->siguiente != nullptr &&
               actual->siguiente->id != id) {
            actual = actual->siguiente;
        }

        // Si encontramos el nodo.
        if (actual->siguiente != nullptr) {

            // Nodo a eliminar.
            Estudiante* aBorrar = actual->siguiente;

            // Saltamos el nodo eliminado.
            actual->siguiente = aBorrar->siguiente;

            // Liberamos memoria.
            delete aBorrar;

            cout << "Estudiante eliminado exitosamente." << endl;
        }
        else {
            // No se encontró el ID.
            cout << "Estudiante con ID " << id
                 << " no encontrado." << endl;
        }
    }

    // GUARDAR EN ARCHIVO
    void guardarEnArchivo(string filename) {
        ofstream archivo(filename); // Abrimos el archivo en modo escritura por eso el ofstream. // Si existe, su contenido se reemplaza.

        // Verificamos si se abrió correctamente.
        if (!archivo.is_open()) {
            cout << "Error al abrir el archivo para guardar." << endl;
            return;
        }

        // Recorremos toda la lista.
        Estudiante* actual = cabeza;

        while (actual != nullptr) {

            // Guardamos los datos separados por '|'.
            // Ejemplo:
            // Forma en que se va a guardar >>>   1|Juan Perez|juan@correo.com|Ingenieria|2023
            archivo << actual->id << "|"
                    << actual->nombre << "|"
                    << actual->correo << "|"
                    << actual->escuela << "|"
                    << actual->anioIngreso << "\n";

            // Avanzamos al siguiente nodo.
            actual = actual->siguiente;
        }

        // Cerramos el archivo.
        archivo.close();

        cout << "Datos guardados en "
             << filename
             << " exitosamente." << endl;
    }

    // CARGAR DESDE ARCHIVO
    void cargarDesdeArchivo(string filename) {

        // Abrimos el archivo para lectura.
        ifstream archivo(filename);

        // Si no existe, simplemente salimos.
        if (!archivo.is_open()) {
            return;
        }

        // Variable para almacenar cada línea del archivo.
        string linea;

        // Guardará el ID máximo encontrado.
        int maxId = 0;

        // Limpiar la lista actual antes de cargar.
        while (cabeza != nullptr) {
            Estudiante* aBorrar = cabeza;
            cabeza = cabeza->siguiente;
            delete aBorrar;
        }

        // Puntero al último nodo para insertar eficientemente.
        Estudiante* ultimo = nullptr;

        // Leemos el archivo línea por línea.
        while (getline(archivo, linea)) {

            // Ignoramos líneas vacías.
            if (linea.empty())
                continue;

            // Convertimos la línea en un flujo.
            stringstream ss(linea);

            // Variable auxiliar para campos numéricos.
            string item;

            // Creamos un nuevo nodo.
            Estudiante* nuevo = new Estudiante;
            nuevo->siguiente = nullptr;

            // Extraemos cada campo separado por '|'.
            getline(ss, item, '|');
            nuevo->id = stoi(item);

            getline(ss, nuevo->nombre, '|');
            getline(ss, nuevo->correo, '|');
            getline(ss, nuevo->escuela, '|');

            getline(ss, item, '|');
            nuevo->anioIngreso = stoi(item);

            // Actualizamos el ID máximo.
            if (nuevo->id > maxId) {
                maxId = nuevo->id;
            }

            // Insertamos el nodo al final.
            if (cabeza == nullptr) {
                cabeza = nuevo;
                ultimo = nuevo;
            }
            else {
                ultimo->siguiente = nuevo;
                ultimo = nuevo;
            }
        }

        // Cerramos el archivo.
        archivo.close();

        // El próximo ID será uno mayor que el máximo encontrado.
        siguienteId = maxId + 1;

        cout << "Datos cargados desde "
             << filename
             << " exitosamente." << endl;
    }
};

// MOSTRAR MENÚ PRINCIPAL
void mostrarMenu() {
    cout << "\n=== SISTEMA DE REGISTRO DE ESTUDIANTES ===" << endl;
    cout << "1. Agregar nuevo estudiante" << endl;
    cout << "2. Mostrar todos los estudiantes" << endl;
    cout << "3. Buscar estudiante por ID o nombre" << endl;
    cout << "4. Modificar datos de un estudiante" << endl;
    cout << "5. Eliminar un estudiante" << endl;
    cout << "6. Guardar los datos en archivo .txt" << endl;
    cout << "7. Cargar los datos desde archivo" << endl;
    cout << "8. Salir" << endl;
    cout << "Seleccione una opcion: ";
}

// FUNCIÓN PRINCIPAL
int main() {

    // Creamos el objeto que administrará la lista.
    ListaEstudiantes lista;

    // Variable para almacenar la opción del menú.
    int opcion;

    // Nombre del archivo donde se guardarán los datos.
    string filename = "registros.txt";

    // Cargamos automáticamente los datos al iniciar el programa.
    lista.cargarDesdeArchivo(filename);

    // Bucle principal del programa.
    do {
        // Mostramos el menú.
        mostrarMenu();

        // Validamos que el usuario ingrese un número.
        if (!(cin >> opcion)) {

            // Limpiamos el estado de error.
            cin.clear();

            // Eliminamos el contenido inválido del buffer.
            cin.ignore(10000, '\n');

            cout << "Entrada invalida. Ingrese un numero." << endl;

            // Reinicia el ciclo.
            continue;
        }

        // Evaluamos la opción elegida.
        switch (opcion) {

            // -----------------------------------------------------------------
            // OPCIÓN 1: AGREGAR ESTUDIANTE
            // -----------------------------------------------------------------
            case 1: {
                string nombre, correo, escuela;
                int anio;

                // Limpiamos buffer antes de getline.
                cin.ignore(10000, '\n');

                cout << "Ingrese nombre completo: ";
                getline(cin, nombre);

                cout << "Ingrese correo electronico: ";
                getline(cin, correo);

                cout << "Ingrese escuela profesional: ";
                getline(cin, escuela);

                cout << "Ingrese anio de ingreso: ";
                cin >> anio;

                // Insertamos el estudiante en la lista.
                lista.agregarEstudiante(nombre, correo,
                                        escuela, anio);
                break;
            }

            // -----------------------------------------------------------------
            // OPCIÓN 2: MOSTRAR ESTUDIANTES
            // -----------------------------------------------------------------
            case 2:
                lista.mostrarEstudiantes();
                break;

            // -----------------------------------------------------------------
            // OPCIÓN 3: BUSCAR
            // -----------------------------------------------------------------
            case 3: {
                int subopcion;

                cout << "1. Buscar por ID\n"
                     << "2. Buscar por Nombre\n"
                     << "Seleccione: ";

                if (!(cin >> subopcion)) {
                    cin.clear();
                    cin.ignore(10000, '\n');
                    cout << "Entrada invalida." << endl;
                    break;
                }

                // Buscar por ID.
                if (subopcion == 1) {
                    int id;
                    cout << "Ingrese ID a buscar: ";
                    cin >> id;
                    lista.buscarEstudiante(id);
                }
                // Buscar por nombre.
                else if (subopcion == 2) {
                    string nombre;
                    cin.ignore(10000, '\n');
                    cout << "Ingrese nombre a buscar: ";
                    getline(cin, nombre);
                    lista.buscarEstudiante(nombre);
                }
                else {
                    cout << "Opcion invalida." << endl;
                }

                break;
            }

            // -----------------------------------------------------------------
            // OPCIÓN 4: MODIFICAR
            // -----------------------------------------------------------------
            case 4: {
                int id;
                cout << "Ingrese el ID del estudiante a modificar: ";
                cin >> id;
                lista.modificarEstudiante(id);
                break;
            }

            // -----------------------------------------------------------------
            // OPCIÓN 5: ELIMINAR
            // -----------------------------------------------------------------
            case 5: {
                int id;
                cout << "Ingrese el ID del estudiante a eliminar: ";
                cin >> id;
                lista.eliminarEstudiante(id);
                break;
            }

            // -----------------------------------------------------------------
            // OPCIÓN 6: GUARDAR
            // -----------------------------------------------------------------
            case 6:
                lista.guardarEnArchivo(filename);
                break;

            // -----------------------------------------------------------------
            // OPCIÓN 7: CARGAR
            // -----------------------------------------------------------------
            case 7:
                lista.cargarDesdeArchivo(filename);
                break;

            // -----------------------------------------------------------------
            // OPCIÓN 8: SALIR
            // -----------------------------------------------------------------
            case 8:
                cout << "Guardando datos antes de salir..." << endl;
                lista.guardarEnArchivo(filename);
                cout << "Saliendo del programa..." << endl;
                break;

            // -----------------------------------------------------------------
            // OPCIÓN INVÁLIDA
            // -----------------------------------------------------------------
            default:
                cout << "Opcion invalida. Intente nuevamente." << endl;
        }

    // El ciclo continúa mientras la opción sea distinta de 8.
    } while (opcion != 8);

    // Fin exitoso del programa.
    return 0;
}

